Add Gauss laser heat source

nmuch opened the pull request:
This PR adds a laser heat source class with a spherical Gauss distribution round the laser center.

The laser volumetric power density is 
<img src="https://latex.codecogs.com/svg.image?p(\boldsymbol{x})&space;=&space;\mathcal{X}_l&space;\,&space;p_{\text{peak}}&space;\exp\left(-2\left(\frac{||\boldsymbol{x}-\boldsymbol{x}_{\text{laser}}||}{r_{\text{laser}}}\right)^2\right)" title="p(x) = \mathcal{X}_l \, p_{\text{peak}} \exp\left(-2\left(\frac{||\boldsymbol{x}-\boldsymbol{x}_{\text{laser}}||}{r_{\text{laser}}}\right)^2\right)" /> 
with the absorptivity <img src="https://latex.codecogs.com/svg.image?\mathcal{X}_l" title="\mathcal{X}_l" /> and a peak power density of
<img src="https://latex.codecogs.com/svg.image?p_{\text{peak}}&space;=&space;\frac{P_{\text{laser}}}{\left(&space;r_{\text{laser}}\,&space;\sqrt{\frac{\pi}{2}}&space;\right)^3}" title="p_{\text{peak}} = \frac{P_{\text{laser}}}{\left( r_{\text{laser}}\, \sqrt{\frac{\pi}{2}} \right)^3}" />
so that the total laser power is
<img src="https://latex.codecogs.com/svg.image?P_{\text{laser}}&space;=&space;\int_{\mathbb{R}^3}&space;p_{\text{peak}}&space;\exp&space;\left(&space;-2&space;\left(\frac{||\boldsymbol{x}&space;-&space;\boldsymbol{x}_{\text{laser}}||}{r_{\text{laser}}}&space;\right)^2&space;\right)&space;d\boldsymbol{x}" title="P_{\text{laser}} = \int_{\mathbb{R}^3} p_{\text{peak}} \exp \left( -2 \left(\frac{||\boldsymbol{x} - \boldsymbol{x}_{\text{laser}}||}{r_{\text{laser}}} \right)^2 \right) d\boldsymbol{x}" />

In order to have different laser heat source models that can be selected via the input file, I introduced a laser heat source interface class.

mschreter commented on include/meltpooldg/heat/laser_heat_source.hpp:
```diff
@@ -0,0 +1,38 @@
+#pragma once
```
Could you please rename the file to `laser_heat_source_base.hpp` and the class to `LaserHeatSourceBase`?

nmuch replied:
Sure.

mschreter commented on include/meltpooldg/heat/laser_heat_source_gauss.hpp:
```diff
@@ -0,0 +1,61 @@
+    double
+    local_compute_volumetric_heat_source(const Point<dim> &position,
+                                         const Point<dim> &laser_position,
+                                         double            power) const override;
```
Here, you could use the keyword `final` instead of override, as long as no other class overrides this member function.

nmuch replied:
I can if you prefer. Usually I don't use `final`.

mschreter commented on include/meltpooldg/heat/laser_heat_source_gauss.hpp:
```diff
@@ -0,0 +1,61 @@
+  /**
+   * Spherical gauss laser heat source.
+   *
+   * local volumetric heat source p:
```
```suggestion
   * local volumetric heat source p (W/m^3):
```

mschreter commented on include/meltpooldg/simulations/melt_front_propagation/melt_front_propagation.hpp:
```diff
@@ -40,27 +42,32 @@ namespace MeltPoolDG::Simulation::MeltFrontPropagation
-    LaserHeatSource(const LaserData<double> &laser_data)
+    LaserHeatSourceFunction(const LaserData<double> &laser_data)
       : Function<dim>()
-      , laser(laser_data.gusarov)
       , center(MeltPoolDG::UtilityFunctions::convert_string_coords_to_point<dim>(laser_data.center))
       , power(laser_data.power)
-
-    {}
+    {
+      if (laser_data.heat_source_model == "Gusarov")
+        laser_model = std::make_unique<Heat::LaserHeatSourceGusarov<dim>>(laser_data.gusarov);
+      else if (laser_data.heat_source_model == "Gauss")
+        laser_model = std::make_unique<Heat::LaserHeatSourceGauss<dim>>(laser_data.gauss);
+      else
+        AssertThrow(false, ExcMessage("Unknown laser heat source model! Abort..."));
+    }
```
I am somehow a bit ambivalent about the concept that the `LaserHeatSource` will be part of the `simulation` file. If we decide that in the `simulation` file the user has to provide a function for the heat source, then we could also outsource the parsing of the laser parameters from the central `parameters.hpp` file to the individual classes and simply read them in the constructor. For the moment consider it as a comment, but we have to think about conceptual design here.

nmuch replied:
I added it to the simulation file to attach the laser heat source with `attach_source_field()`. Here you have the most flexibility to modify the lasers behavior since for every simulation you can write a custom `Function`.

mschreter commented on include/meltpooldg/heat/laser_heat_source.hpp:
```diff
@@ -0,0 +1,38 @@
+    /**
+     * Compute a DoF vector of the heat source.
+     */
+    virtual void
```
Should this function also be virtual?

peterrum replied:
If I see it correctly, the `virtual` keyword can be removed. `compute_volumetric_heat_source()` is implemented in the base class and only `local_compute_volumetric_heat_source()` is implemented in the derived classes. One could consider to add a `final` here ;)

mschreter commented on include/meltpooldg/heat/laser_heat_source_gauss.hpp:
```diff
@@ -0,0 +1,61 @@
+   * p = absorptivity * p_peak * exp( -2 ||x - laser_center||^2 / laser_radius^2 ) [1]
+   *
+   * with peak laser power density p_peak:
+   * p_peak = laser_power / ( laser_radius * sqrt(pi/2) )^3
```
Do you also have a reference for this equation?

nmuch replied:
No. In the SPH paper, the peak laser power density <img src="https://latex.codecogs.com/svg.image?p_{\text{peak}}" title="p_{\text{peak}}" /> (in the paper: <img src="https://latex.codecogs.com/svg.image?s_{l0}^{lg}" title="s_{l0}^{lg}" />) is assumed to be known. However, we input the absolute laser power <img src="https://latex.codecogs.com/svg.image?P_{\text{laser}}" title="P_{\text{laser}}" />. 

With this equation I made sure that the total laser power is equal to the integral of the laser power density (before absorptivity). See the third equation of the PR description.

mschreter reviewed (approved):
Thanks @nmuch. Is there a typo in the last equation of your PR description, where also the absorptivity should be included or do I misunderstand something here? Anyway, some comments from my side, but overall I am happy with the results :-).

nmuch commented:
Tnank you @mschreter for your review. I'll make the obvious changes.

>  Is there a typo in the last equation of your PR description, where also the absorptivity should be included or do I misunderstand something here?

Imo this equation should not include the absorptivity because the laser energy that is not absorbed is part of the total laser power. If not, we can disregard the absorptivity all together.

mschreter commented:
Thanks @nmuch. Once CI is happy I'll merge.

mschreter commented:
@nmuch: One general question -- what do you mean by adding "Draft:" in the title of this PR :-)? This does not mean that this PR is still work in progress, does it? Thanks for clarification.

nmuch commented:
> what do you mean by adding "Draft:" in the title of this PR :-)?

@mschreter Oh this is a habit from using Gitlab in other projects. I added draft because I was not 100% sure the equation I introduced were correct when creating this PR and because of #207. I can remove the Draft: now I think.

mschreter commented:
> > what do you mean by adding "Draft:" in the title of this PR :-)?
> 
> @mschreter Oh this is a habit from using Gitlab in other projects. I added draft because I was not 100% sure the equation I introduced were correct when creating this PR and because of #207. I can remove the Draft: now I think.

I see. Let me summarize my experiences regarding #207 in the corresponding issue.
